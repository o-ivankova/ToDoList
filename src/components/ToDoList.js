import React from 'react';
import ToDoListItem from './ToDoListItem';

const ToDoList = () => {
    return (
        <ul>
        <li><ToDoListItem content='Drink coffee'/></li>
        <li><ToDoListItem
        content='Create an app'
        important/></li>
        </ul>
    );
};

export default ToDoList;