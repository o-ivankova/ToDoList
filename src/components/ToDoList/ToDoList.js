import React from 'react';
import ToDoListItem from '../ToDoListItem';
import './ToDoList.scss';
//import PropTypes from 'prop-types';


/* ToDoList.propTypes = {
    elements: PropTypes.array,
  }; */

// eslint-disable-next-line react/prop-types
const ToDoList = ( { elements } ) => {

    // eslint-disable-next-line react/prop-types
    const list = elements.map((item) => {
     const { id, ...itemProps } = item;
      return (
        <li className='list-group-item' key={ id }>
        <ToDoListItem { ...itemProps } />
       </li>)
       });

    return (
        <ul className='list-group'>
        { list }
        </ul>
    );
};

export default ToDoList;