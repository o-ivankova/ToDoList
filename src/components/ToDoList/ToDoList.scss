.list .list-group-item {
    padding: .25rem .75rem;
    margin: 4px 4px 4px;
    display: grid;
    gap: 2px;
    justify-content: left;
  }