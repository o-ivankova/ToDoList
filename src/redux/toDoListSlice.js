import { createSlice } from "@reduxjs/toolkit";
import storage from "redux-persist/lib/storage";
import { persistReducer } from "redux-persist";

const initialState = {
  elements: [],
  elementId: 0,
  filter: "all",
};

export const toDoListSlice = createSlice({
  name: "toDoList",
  initialState,
  reducers: {
    addElement: (state, action) => {
      const input = action.payload;

      if (input !== "") {
        state.elementId += 1;

        const newElement = {
          content: input,
          completed: false,
          id: state.elementId,
        };

        state.elements.push(newElement);
      }
    },

    deleteElement: (state, action) => {
      const id = action.payload;
      state.elements = state.elements.filter((el) => el.id !== id);
    },

    deleteAllCompleted: (state) => {
      state.elements = state.elements.filter((el) => !el.completed);
    },

    checkCompleted: (state, action) => {
      const id = action.payload;
      const index = state.elements.findIndex((el) => el.id === id);
      state.elements[index].completed = !state.elements[index].completed;
    },

    changeFilter: (state, action) => {
      const filter = action.payload;
      state.filter = filter;
    },

    reorderElements: (state, action) => {
      const result = action.payload;
      if (!result.destination) return;

      const [reorderedEl] = state.elements.splice(result.source.index, 1);
      state.elements.splice(result.destination.index, 0, reorderedEl);
    },
  },
});

export const {
  addElement,
  deleteElement,
  checkCompleted,
  deleteAllCompleted,
  changeFilter,
  reorderElements,
} = toDoListSlice.actions;

const persistConfig = {
  key: "root",
  storage,
};

const { reducer } = toDoListSlice;

export const persistedReducer = persistReducer(persistConfig, reducer);

// export default reducer;
