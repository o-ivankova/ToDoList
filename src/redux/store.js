import { configureStore } from "@reduxjs/toolkit";
import {persistedReducer} from "./toDoListSlice";
import { persistStore } from 'redux-persist'

export const store = configureStore({
  reducer: {
    toDoList: persistedReducer,
  },
});

export const persistor = persistStore(store);