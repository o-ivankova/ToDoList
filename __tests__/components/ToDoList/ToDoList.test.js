import React from "react";
import "@testing-library/jest-dom";
import { jest, expect, test, describe, afterAll, afterEach } from "@jest/globals";
import { fireEvent, render, screen } from "@testing-library/react";
import ToDoList from "../../../src/components/ToDoList/ToDoList";
import { Provider } from "react-redux";
import { setupStore } from "../../../src/redux/store";
import renderer from "react-test-renderer";

describe("ToDoList tests", () => {
  test("component is rendered", () => {
    const preloadedState = {
      toDoList: {
        filter: "all",
        elements: [],
      },
    };
    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <ToDoList />
      </Provider>
    );

    expect(screen.getByTestId("list")).toBeInTheDocument();
  });

  test("ToDoListItems are rendered", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: true },
          { id: 2, content: "test2", completed: true },
          { id: 3, content: "test3", completed: false },
          { id: 4, content: "test4", completed: false },
        ],
      },
    };
    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <ToDoList />
      </Provider>
    );

    expect(screen.findByText("test1")).not.toBeNull;
    expect(screen.findByText("test2")).not.toBeNull;
    expect(screen.findByText("test3")).not.toBeNull;
    expect(screen.findByText("test4")).not.toBeNull;
  });

  test("checkbox click", () => {
    const preloadedState = {
      toDoList: {
        elements: [{ id: 1, content: "test1", completed: false }],
      },
    };
    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <ToDoList />
      </Provider>
    );

    fireEvent.click(screen.getByTestId("checkbox"));

    const elementsInLocalStorage = JSON.parse(
      localStorage.getItem("toDoElements")
    );

    expect(elementsInLocalStorage[0].completed).toBe(true);
    expect(screen.getByTestId("checkbox").checked).toBe(true);
    expect(screen.getByTestId("to-do-list-item")).toHaveClass("todoListItemCompleted");
  });

  test("filter change", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: false },
          { id: 2, content: "test2", completed: true },
        ],
        filter: "completed",
      },
    };
    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <ToDoList />
      </Provider>
    );

    expect(screen.queryByText("test1")).not.toBeInTheDocument();
    expect(screen.getByText("test2")).toBeInTheDocument();
  });

  test("snapshot test", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: false },
          { id: 2, content: "test2", completed: true },
        ],
        filter: "completed",
      },
    };
    const mockStore = setupStore(preloadedState);

    const domTree = renderer
      .create(
        <Provider store={mockStore}>
        <ToDoList />
      </Provider>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    localStorage.setItem("toDoElements", JSON.stringify([]));
  });
});
