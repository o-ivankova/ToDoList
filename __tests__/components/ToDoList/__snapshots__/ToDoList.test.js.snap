// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList tests snapshot test 1`] = `
<ul
  className="list"
  data-rbd-droppable-context-id="4"
  data-rbd-droppable-id="list"
  data-testid="list"
>
  <li
    aria-describedby="rbd-hidden-text-4-hidden-text-14"
    className="list-group-item"
    data-rbd-drag-handle-context-id="4"
    data-rbd-drag-handle-draggable-id="2"
    data-rbd-draggable-context-id="4"
    data-rbd-draggable-id="2"
    data-testid="list-group-item"
    draggable={false}
    onDragStart={[Function]}
    onTransitionEnd={null}
    role="button"
    style={
      Object {
        "transform": null,
        "transition": null,
      }
    }
    tabIndex={0}
  >
    <span
      className="todoListItemCompleted"
      data-testid="to-do-list-item"
    >
      <div
        className="todoListItemContent"
        data-testid="todo-list-item-content"
      >
        <input
          checked={true}
          className="checkbox"
          data-testid="checkbox"
          onChange={[Function]}
          type="checkbox"
        />
        test2
      </div>
      <button
        className="deleteBtn"
        data-testid="delete-btn"
        onClick={[Function]}
        type="button"
      >
        <svg
          data-testid="delete-icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708z"
          />
        </svg>
      </button>
    </span>
  </li>
</ul>
`;
